### PR TITLE
[do-not-review] [loss] Scope chunked FSDP finalization to a training step

### PR DESCRIPTION
**Pull request by xmfan**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* (to be filled)

## Why

`ChunkedLossWrapper` keeps an FSDP-wrapped language-model head unsharded while it processes the chunks of one loss call. A pipeline step invokes the loss once per target microbatch, so finalizing FSDP after every call repeatedly reshards the same weight and enables gradient synchronization before the step is complete.

The ownership boundary is the complete pipeline training step, not an individual microbatch loss call.

## FSDP lifetime

```text
Before

loss(microbatch 0): unshard -> chunks -> enable grad sync -> reshard
loss(microbatch 1): unshard -> chunks -> enable grad sync -> reshard
...

After

step_context(num_loss_calls=N)
    |
    +--> loss 0: unshard -> chunks -> keep unsharded
    +--> loss 1:             chunks -> keep unsharded
    +--> ...
    +--> loss N-1 final chunk -> enable grad sync -> reshard
```

## Change

Add an exception-safe `BaseLoss.step_context()` with an exact loss-call count. The chunked implementation keeps the language-model head unsharded across all declared calls, restores gradient synchronization only for the final chunk of the final call, and restores FSDP policy and resharding on both success and failure.

Reject nested contexts and extra or missing loss calls instead of carrying invalid state into later steps.

Dynamic pipeline initialization may invoke the loss once to infer backward metadata, so the first eager schedule invocation remains unscoped. Validation invalidates that training metadata. GraphPP skips subsequent Python loss scopes because it replays the traced loss graph without calling the Python wrapper.

## Testing

The tests cover multi-call finalization, exception cleanup, exact call counts, pipeline initialization, validation invalidation, and GraphTrainer behavior.

- `python -m pytest -q tests/unit_tests/cpu/test_loss.py tests/unit_tests/cpu/test_trainer.py torchtitan/experiments/graph_trainer/tests/test_trainer.py`
- Eight-GPU `validation_tp_cp_pp` integration, with validation at steps 5 and 10 and 10/10 completed training steps

stack-info: PR: https://github.com/pytorch/torchtitan/pull/4431, branch: xmfan/stack/48

### CODE REVIEW

**Comment by xmfan** [comment]
Closing duplicate PR created by an incorrect ghstack resubmission; the original stack remains in #4429-#4436.
